renn/utils.py

```diff
@@ -137,3 +139,85 @@ def one_hot(labels, num_classes, dtype=jnp.float32):
     one_hot_labels: array with shape (num_examples, num_classes).
   """
   return jnp.array(jnp.array(labels)[:, None] == jnp.arange(num_classes), dtype)
+
+
+def select_output(sequences, indices):
+  """Given an array of shape (batch_size, sequence_length, num_outputs),
+  and a 1D array specifying which indices of each sequence to select, return
+  a (batch_size, num_ouputs)-shaped array with the selected elements.
+
+  Args:
+    sequences: array with shape (batch_size, sequence_length, num_outputs)
+    indices: 1D array with length batch_size
+
+  Returns:
+    selected_outputs: array with shape (batch_size, num_outputs)
+  """
+
+  assert len(indices) == sequences.shape[0]
+
+  # shape indices properly
+  indices_shaped = indices[:, jnp.newaxis, jnp.newaxis]
+
+  # select element
+  selected_outputs = jnp.take_along_axis(sequences, indices_shaped, axis=1)
+
+  # remove sequence dimension
+  selected_outputs = jnp.squeeze(selected_outputs, axis=1)
+
+  return selected_outputs
+
+def make_loss_function(network_apply_fun, loss_type = 'xent', num_outputs = 1):
+  """ Given a network function, return a loss function which maps a tuple of
```

Comment:
Suggestion for a cleaner API: just have the loss function be an argument that is directly passed as an argument.

That simplifies the number of arguments this function needs (instead of having both a `loss_type` and a `num_outputs`, it just has a single argument, the function). In addition, as we add potentially new loss functions (such as MSE) we do not need to edit/update this function.

Reply:
Agree, made this change